### PR TITLE
[lite-version] Separate into melange.ppx and melange.ppx-lib

### PR DESCRIPTION
**Pull request by anmonteiro**
- decouples a `melange.ppx` from the main melange executable
  - the main goal of this change is to fix https://github.com/ocaml/dune/pull/7323
    - [x] TODO: include that test in this repo
  - the key insight here is that, in melange, only `%bs.obj` etc. extensions can potentially introduce new dependencies for ocamldep, so those are the transformations that the PPX runs
- we ran into a few problems with the approach in #527 due to the exhaustiveness checking for `[@bs.*]` attributes
- this approach migrates the extensions first, to ppxlib context-free rules, which are order independent
- it also sets us up to migrate more transformations incrementally to the `melange.ppx`

in summary:
- we fix the bug that prompted this change
- we set up a path towards incremental migration without having to do it all at once

part of #463

### CODE REVIEW

**Comment by anmonteiro** [comment]
This is still not 100% right. `Ast_derive_js_mapper` emits extensions which are expected to be expanded later, but will now need to be expanded inline.

File: reactjs-jsx-ppx/reactjs_jsx_ppx_v3.ml [hunk 1]
```diff
@@ -1315,5 +1312,5 @@ let jsxMapper =
 
 let () =
   Driver.V2.register_transformation "reactjs"
-    ~impl:(fun ctxt str -> jsxMapper#structure ctxt str)
-    ~intf:(fun ctxt sig_ -> jsxMapper#signature ctxt sig_)
+    ~preprocess_impl:(fun ctxt str -> jsxMapper#structure ctxt str)
+    ~preprocess_intf:(fun ctxt sig_ -> jsxMapper#signature ctxt sig_)
```
Comment (jchavarri) [inline reactjs-jsx-ppx/reactjs_jsx_ppx_v3.ml:1316]: Why does react ppx need to change to `preprocess_*` arguments? Won't this play badly with other ppxs?

Comment (anmonteiro) [reply reactjs-jsx-ppx/reactjs_jsx_ppx_v3.ml:1316]: indeed. this needs to be reverted. I was just trying things out.

**Comment by anmonteiro** [comment]
The last few commits allow making progress here without having https://github.com/ocaml/dune/pull/7373 land first.
